ghproxy/ghcache/ghcache.go

```diff
@@ -138,18 +139,122 @@ func cacheResponseMode(headers http.Header) CacheResponseMode {
 	return ModeMiss
 }
 
-func newThrottlingTransport(maxConcurrency int, roundTripper http.RoundTripper) http.RoundTripper {
-	return &throttlingTransport{sem: semaphore.NewWeighted(int64(maxConcurrency)), roundTripper: roundTripper}
+func newThrottlingTransport(maxConcurrency int, requestThrottlingTime, getRequestThrottlingTime uint, roundTripper http.RoundTripper, hasher ghmetrics.Hasher) http.RoundTripper {
+	return &throttlingTransport{
+		sem:                      semaphore.NewWeighted(int64(maxConcurrency)),
+		roundTripper:             roundTripper,
+		requestThrottlingTime:    requestThrottlingTime,
+		getRequestThrottlingTime: getRequestThrottlingTime,
+		hasher:                   hasher,
+		apiv3:                    newTokensRegistry(requestThrottlingTime, getRequestThrottlingTime),
+		apiv4:                    newTokensRegistry(requestThrottlingTime, getRequestThrottlingTime),
+	}
+}
+
+func newTokensRegistry(requestThrottlingTime, getRequestThrottlingTime uint) tokensRegistry {
+	return tokensRegistry{
+		lock:              sync.Mutex{},
+		tokens:            map[string]tokenInfo{},
+		throttlingTime:    time.Millisecond * time.Duration(requestThrottlingTime),
+		getThrottlingTime: time.Millisecond * time.Duration(getRequestThrottlingTime),
+	}
+}
+
+// tokenInfo keeps the last request timestamp and information whether it was GET request
+type tokenInfo struct {
+	getReq    bool
+	timestamp time.Time
+}
+
+// tokenRegistry keeps the timestamp of last handled request per token budget (appId or hash)
+type tokensRegistry struct {
+	lock              sync.Mutex
+	tokens            map[string]tokenInfo
+	throttlingTime    time.Duration
+	getThrottlingTime time.Duration
+}
+
+func (tr *tokensRegistry) getRequestWaitDuration(tokenBudgetName string, getReq bool) time.Duration {
+	var duration time.Duration
+	tr.lock.Lock()
+	defer tr.lock.Unlock()
+	toQueue := time.Now()
+	if t, exists := tr.tokens[tokenBudgetName]; exists {
+		toQueue, duration = tr.calculateRequestWaitDuration(t, toQueue, getReq)
+	}
+	tr.tokens[tokenBudgetName] = tokenInfo{getReq: getReq, timestamp: toQueue}
+	return duration
+}
+
+func (tr *tokensRegistry) calculateRequestWaitDuration(t tokenInfo, toQueue time.Time, getReq bool) (time.Time, time.Duration) {
+	throttlingTime := tr.throttlingTime
+	followingGetRequests := t.getReq && getReq
```

Comment:
The identifiers `t` and `tr` tend to mix with each other, I suggest to rename `t` to `lastRequest` (it is much more an info about the last request than anything related to tokens)